src/Microsoft.ML.Transforms/SvmLight/SvmLightLoader.cs

```diff
@@ -0,0 +1,851 @@
+    /// having any status.
+    ///
+    /// Labels are floating point values for regression, and for binary classification being one
+    /// of {-1, 0, 1/+1}. Negative class examples are -1, positive class examples are +1, and 0
+    /// indicates that the label is unknown and should be classified using semi-supervised
+    /// techniques.
+    ///
+    /// The "0 label" semi-supervised capability was rarely used and none of our learners
+    /// currently do anything like this, though it is possible we may introduce semi-supervised
+    /// learners in the future. For now this loader just parses this as a single floating point
+    /// value, period. (Which means, for our binary classifier learners, that 0 and -1 will both
+    /// be treated identically.) If we were to support this, the natural thing would be to have
+    /// an option to map 0 to NA, somehow. But practically, variants of the SVM-light format have
+    /// promulgated to the point where nearly every time 0 is used, it actually refers to a
+    /// negative example, so we may continue to accept this corruption as "correct" by default.
+    ///
+    /// The actual feature vector is specified through a series of key/value pairs. SVM-light
+    /// requires that the keys be positive, increasing integers, except for three special keys:
+    /// cost (we interpret as Weight), qid (we interpret as GroupId) and sid (we ignore these,
+    /// but might present them as a column in the future if any of our learners implement anything
+    /// resembling slack id). The value for 'cost' is float, 'qid' is a long, and 'sid' is a long
+    /// that must be positive. If these keys are specified multiple times, the last one wins.
+    ///
+    /// SVM-light, if the tail of the value is not interpretable as a number, will ignore the tail.
+    /// E.g., "5:3.14hello" will be interpreted the same as "5:3.14". I am aware of one real dataset
+    /// that took advantage of this, and for now I do not support this type of thing.
+    ///
+    /// We do not retain the restriction on keys needing to be increasing values in our loader,
+    /// due to the way we compose our feature vectors, but it will be most efficient if this policy
+    /// is still followed. If it is followed a sort will not be required.
+    ///
+    /// This loader has the special option through the <c>xf</c> option to specify a transform,
+    /// possibly trainable, to convert the raw text of the key values into the key value. The
+    /// transform, whatever it is, must in addition to user specified options accept an argument
+    /// of the form "column=Name" to identify a column to convert. Ideally there would be some
+    /// other way to specify this other than hacking arguments. The intent of this is to allow
+    /// things like string keys, a common variant of the format, but one emphatically not allowed
+    /// by the original format.
```

Comment:
This variation existed in the internal version of this loader, but I have not been able to find any references to it online. How important do you think it is to support this?

Reply:
As we discussed on the call, I think it is important to support this. Since SvmLightLoader is implemented as data loader, it is typically the first stage in the pipeline. If someone has data in this format, they will not be able to load the data into ML.NET if we don't support this.

---
In reply to: [322014712](https://github.com/dotnet/machinelearning/pull/4190#discussion_r322014712) [](ancestors = 322014712)